File: bindings/ruby/examples/box512.rb
# frozen_string_literal: true

# crypto_box512 (l(p)=512 sibling of crypto_box, T-193/T-204): generate a keypair, seal a message
# to the public key, open it with the secret key.
#
# Run: ruby examples/box512.rb

$LOAD_PATH.unshift(File.expand_path("../lib", __dir__))
require "dstu_core"

secret_key = DstuCore.box512_keygen
public_key = DstuCore.box512_public_key(secret_key) # safe to share/publish

message = "a message for the public key's holder only"
sealed = DstuCore.box512_seal(public_key, message)
opened = DstuCore.box512_open(secret_key, sealed)
raise "round trip failed" unless opened == message

puts "sealed #{message.bytesize} bytes -> #{sealed.bytesize} bytes, round-tripped OK"

tampered = sealed.dup
tampered[-1] = (tampered[-1].ord ^ 1).chr
begin
  DstuCore.box512_open(secret_key, tampered)
rescue DstuCore::Error
  puts "tampered ciphertext correctly rejected"
end

File: bindings/php/examples/box512.php
<?php

declare(strict_types=1);

/**
 * crypto_box512 (l(p)=512 sibling of crypto_box, T-193/T-204): generate a keypair, seal a message
 * to the public key, open it with the secret key.
 *
 * Run: php -d extension=<path to dstu_core_php.dll/.so> examples/box512.php
 */

$secretKey = dstu_core_box512_keygen();
$publicKey = dstu_core_box512_public_key($secretKey); // safe to share/publish

$message = "a message for the public key's holder only";
$sealed = dstu_core_box512_seal($publicKey, $message);
$opened = dstu_core_box512_open($secretKey, $sealed);
if ($opened !== $message) {
    throw new RuntimeException('round trip failed');
}

printf("sealed %d bytes -> %d bytes, round-tripped OK\n", strlen($opened), strlen($sealed));

$tampered = $sealed;
$tampered[strlen($tampered) - 1] = chr(ord($tampered[-1]) ^ 1);
try {
    dstu_core_box512_open($secretKey, $tampered);
} catch (DstuCoreException $e) {
    echo "tampered ciphertext correctly rejected\n";
}
